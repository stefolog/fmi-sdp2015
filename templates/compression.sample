A
ABCD
2(A)
2(AB)
AB2(C)
AB13(C)
3(X2(B)Y)
2(A3(B4(C)))
0(A)
1(A)
