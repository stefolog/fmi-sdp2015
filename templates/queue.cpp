#include <iostream>

using namespace std;

template <class T>
struct elem_q {
  T inf;
  elem_q<T>* link;
};

template <class T>
class queue
{
  public:
    queue();
    ~queue();
    queue(queue const &);
    queue<T>& operator=(queue const &);
    void push(T const &);
    int pop(T &);
    void head(T &);
    void print();
    bool empty() const;

  private:
    elem_q<T> *front, *rear;
    void delqueue();
    void copy(queue const &);
};

template <class T>
queue<T>::queue() {
  front = rear = NULL;
}

template <class T>
queue<T>::~queue() {
  delqueue();
}

template <class T>
queue<T>::queue(queue const & r) {
  front = rear = NULL;
  copy(r);
}

template <class T>
queue<T>& queue<T>::operator=(queue const & r) {
  if (this != &r) {
    delqueue();
    copy(r);
  }
  return *this;
}

template <class T>
void queue<T>::push(T const & x) {
  elem_q<T> * e =  new elem_q<T>;
  e->inf = x;
  e->link = NULL;

  if (rear) {
    rear->link = e;
    rear = e;
  } else {
    front = rear = e;
  }
}

template <class T>
void queue<T>::head(T & x) {
  if (!rear) {
    throw 1;
  }

  x = front->inf;
}

template <class T>
int queue<T>::pop(T & x) {
  if (!rear) {
    return 0;
  }

  if (front == rear) {
    rear = NULL;
  }

  elem_q<T> * temp = front;
  front = front->link;

  x = temp->inf;
  delete temp;

  return 1;
}

template <class T>
void queue<T>::print() {
  T x;
  while (pop(x)) {
    cout << x << " ";
  }
  cout << endl;
}

template <class T>
bool queue<T>::empty() const {
  return rear == NULL;
}

template <class T>
void queue<T>::delqueue() {
  elem_q<T>* temp;
  while (front) {
    temp = front;
    front = front->link;
    delete temp;
  }

  front = rear = NULL;
}

template <class T>
void queue<T>::copy(queue const & r) {
  elem_q<T> *p = r.front;
  while(p) {
    push(p->inf);
    p = p->link;
  }
}
