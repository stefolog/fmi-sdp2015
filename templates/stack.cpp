#include <iostream>

using namespace std;

template <class T>
class stack;

template <class T>
class item {
  friend class stack<T>;
  private:
  item(T x = 0) {
    inf = x;
    link = NULL;
  }

  T inf;
  item* link;
};


template <class T>
class stack {
  public:
  stack(T x);
  stack();
  ~stack();
  stack(stack const &);

  stack<T>& operator=(stack const &);

  void push(T const &);
  void pop(T&);
  void top(T&) const;
  bool empty() const;
  void print();

  private:
  item<T>* start;
  void delstack();
  void copy(stack const &);
};

// Constructors & operators
template <class T>
stack<T>::stack(T x) {
  start = new item<T>(x);
}

template <class T>
stack<T>::stack() {
  start = NULL;
}

template <class T>
stack<T>::stack(stack const & r) {
  copy(r);
}

template <class T>
stack<T>::~stack() {
  delstack();
}

template <class T>
stack<T>& stack<T>::operator=(stack<T> const & r)  {
  if (this != &r) {
    delstack();
    copy(r);
  }
  return *this;
}

// Stack API
template <class T>
void stack<T>::push(T const & x) {
  item<T> * i = new item<T>(x);
  i->link = start;
  start = i;
}

template <class T>
void stack<T>::pop(T & x) {
  if (!start) {
    throw 1;
  }

  item<T>* temp = start;
  start = start->link;

  x = temp->inf;
  delete temp;
}

template <class T>
void stack<T>::top(T& res) const {
  res = start->inf;
}

template <class T>
bool stack<T>::empty() const {
  return start == NULL;
}

template <class T>
void stack<T>::print() {
  T x;
  while (!empty()) {
    pop(x);
    cout << x << " ";
  }
  cout << endl;
}

// Private helpers
template <class T>
void stack<T>::delstack() {
  item<T> * p;
  while (start) {
    p = start;
    start = start->link;
    delete p;
  }
}

template <class T>
void stack<T>::copy(stack const & r) {
  if (r.start) {
    item<T> *p = r.start;
    item<T> *q = NULL;

    start = new item<T>;
    start->inf = p->inf;
    start->link = q;

    q = start;
    p = p->link;
    while (p) {
      q->link = new item<T>;
      q->link->inf = p->inf;
      q->link->link = NULL;

      q = q->link;
      p = p->link;
    }
  } else {
    start = NULL;
  }
}
