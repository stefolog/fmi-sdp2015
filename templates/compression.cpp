#include <iostream>
#include <cstring>
#include <fstream>
#include "stack.cpp"
#include "queue.cpp"

using namespace std;

// A2(BC)
//  f^
// cout << A
// repeat 2, expr [B, C]
//      decompress(f -> , expr = [] )
// get(c)
// A2(BC)
//   f^     expr = [B]
// A2(BC)
//    f^     expr = [B, C]


// A2(BC3(X))
//   f^
// cout << A
// A2(BC3(X))
//  f^ repeat 2, expr [ ], decompress()
//     expr [B, C]
//     A2(BC3(X))
//         f^
//        --repeat 3, expr [ ], decompress()
//        | A2(BC3(X))
//        |       f^
//        |        expr [X]
//        +->repeat 3, expr [X]
//     expr [B, C, X ,X, X]

// 3(X2(B)Y)
//  ^
//   get -> X
// A-Z -> expr
// 2(A-Z) -> expr(A-Z, A-Z)
// ) return
void decompress(ifstream& input, queue<char>& expr) {
  while (input) {
    char c;
    input.get(c);
    if (')' == c) {
      return;

    } else if (c == '\n') {
      expr.push(c);

    } else if ('A' <= c && c <= 'Z') {
      expr.push(c);

    } else if ('0' <= c && c <= '9') {
      int repeat = c - '0';
      while (true) {
        input.get(c);
        if (c == '(') {
          break;
        }
        repeat *= 10;
        repeat += c - '0';
      }

      queue<char> internalRepeat;
      decompress(input, internalRepeat);

      while (repeat--) {
        queue<char> temp = internalRepeat;
        while (!temp.empty()) {
          char repeatChar;
          temp.pop(repeatChar);
          expr.push(repeatChar);
        }
      }
    }
  }
}

void printFile(const char * filename) {
  ifstream input(filename);
  while (input) {
    char c;
    input.get(c);
    cout << c;
  }
  input.close();
}

int main()
{
  // decompress("compression.sample");
  queue<char> expr;
  ifstream input("compression.sample");
  decompress(input, expr);

  while (!expr.empty()) {
    char c;
    expr.pop(c);
    cout << c;
  }

  cout << "EXPECTED =========================" << endl << endl;
  printFile("compression.expected");
  return 0;
}
