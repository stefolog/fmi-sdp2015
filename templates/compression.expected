A
ABCD
AA
ABAB
ABCC
ABCCCCCCCCCCCCC
XBBYXBBYXBBY
ABCCCCBCCCCBCCCCABCCCCBCCCCBCCCC

A
